sim-lib/src/lnd.rs

```diff
@@ -101,12 +104,72 @@ impl LightningNode for LndNode {
         Ok(payment_hash)
     }
 
-    async fn track_payment(&mut self, _hash: PaymentHash) -> Result<(), LightningError> {
-        unimplemented!()
+    async fn track_payment(
+        &mut self,
+        hash: PaymentHash,
+        shutdown: Listener,
+    ) -> Result<PaymentOutcome, LightningError> {
+        let response = match self
+            .client
+            .router()
+            .track_payment_v2(TrackPaymentRequest {
+                payment_hash: hash.0.to_vec(),
+                no_inflight_updates: false,
+            })
+            .await
+        {
+            Ok(response) => response,
+            Err(err) => {
+                return Err(LightningError::TrackPaymentError(err.to_string()));
+            }
+        };
+
+        let mut stream = response.into_inner();
+
+        tokio::select! {
+            biased;
+            stream = stream.message() => {
+                let payment  = match stream {
+                    Ok(payment) => payment,
+                    Err(err) => {
+                        return Err(LightningError::TrackPaymentError(err.to_string()));
+                    }
+                };
+
+                match payment {
+                    Some(payment) => {
+
+                        let htlc_count = payment.htlcs.len();
+                        let settled = payment.status == 2 // routerrpc::PaymentStatus::SUCCEEDED
+                        || payment.status == 3; // routerrpc::PaymentStatus::FAILED
+
+                        let failure_reason = match payment.failure_reason {
+                            1 => Some(PaymentFailureReason::PaymentExpired),
+                            2 => Some(PaymentFailureReason::RouteNotFound),
+                            3 => Some(PaymentFailureReason::UnexpectedError),
+                            4 => Some(PaymentFailureReason::UnexpectedError), // TODO: Capture failure reasons for incorrect payment
+                            5 => Some(PaymentFailureReason::UnexpectedError), // TODO: Capture failure reasons for insufficient balance
+                            _ => None,
+                        };
+                        return Ok(PaymentOutcome::Resolved(PaymentResult {
+                            settled,
+                            htlc_count,
+                            failure_reason,
+                        }));
```

Comment:
I think all of this can be flattened to the `PaymentResult` constructor. There is no need of temp variables